File: packages/core/types/prompt-quality.ts
/**
 * Prompt quality dashboard types (RUYI-184, self-evolution phase 2).
 *
 * The seven dimensions are read from the platform's own run stream; the server
 * folds each day into counts and never divides, so a dimension arrives here as
 * one of three states rather than as a number that might be a stand-in for
 * "nothing was measured". The UI renders the state — it has no branch that can
 * turn a missing measurement into 0%.
 *
 * `state` stays a plain string on the wire type because it is server-driven: a
 * state a newer backend adds must still parse, with the view taking its default
 * branch. The narrowing to a discriminated union happens in
 * `core/self-evolution/measure.ts`, which is the one place that decides what an
 * unknown state renders as.
 */

/** Which prompt tier a dashboard reads. Matches the server's scope segment. */
export type PromptQualityScope = "workspace" | "project" | "squad" | "agent";

/** The seven cards, keyed as the API returns them. */
export type PromptQualityDimension =
  | "injected_tokens"
  | "run_tokens_median"
  | "discipline"
  | "tool_failure_rate"
  | "retry_rate"
  | "failure_attribution"
  | "first_pass_rate";

/** One dimension as the API returns it. */
export interface PromptQualityMeasure {
  /** "ok" | "no_data" | "insufficient_sample", server-driven. */
  state: string;
  /** The rate or absolute figure. Null unless state is "ok". */
  value: number | null;
  numerator: number | null;
  denominator: number | null;
  /** The dimension's own denominator, which is not always the run count. */
  sample: number;
  /** The declared floor the sample was compared against. */
  threshold: number;
  /** Names why the state is not ok. A key, not a sentence. */
  reason?: string;
  /** D6 only: failures dropped from the attribution denominator. */
  excluded?: number;
}

export interface PromptQualityMeasures {
  injected_tokens: PromptQualityMeasure;
  run_tokens_median: PromptQualityMeasure;
  discipline: PromptQualityMeasure;
  tool_failure_rate: PromptQualityMeasure;
  retry_rate: PromptQualityMeasure;
  failure_attribution: PromptQualityMeasure;
  first_pass_rate: PromptQualityMeasure;
}

/** One version's seven cards plus the span they cover. */
export interface PromptQualityVersionMeasures {
  /** 0 for the window aggregate across versions. */
  version: number;
  /** How many daily buckets went into this group. */
  days: number;
  first_day?: string;
  last_day?: string;
  runs: number;
  measures: PromptQualityMeasures;
  /** D6 breakdown by failure reason. Empty when the blob did not parse. */
  failure_reasons: Record<string, number>;
  excluded_failed_runs: number;
}

/** One evidence entry behind a D3 sub-dimension score. Locations, never text. */
export interface PromptPerplexityEvidence {
  tier: string;
  section: string;
  conflicts_with?: string;
  note?: string;
}

/** One D3 sub-dimension. */
export interface PromptPerplexityItem {
  key: string;
  score: number;
  justification: string;
  evidence?: PromptPerplexityEvidence[];
}

/**
 * One D3 score. There is one per runtime profile per version and they are
 * never merged: the profiles are scored against different assembled documents,
 * so an average would describe a prompt no run ever receives (Owner Q17).
 */
export interface PromptQualityPerplexity {
  version: number;
  /** "member" | "leader_task", server-driven. */
  runtime_profile: string;
  /** "low" | "medium" | "high", server-driven. */
  band: string;
  percent_low: number | null;
  percent_high: number | null;
  evidence: PromptPerplexityItem[];
  model: string;
  scored_at: string;
}

/**
 * One data source's state. Only the platform's own tables are `required`;
 * everything else degrades to a footer label, and no card may depend on it.
 */
export interface PromptQualitySource {
  /** "platform" | "scoring_model" | "langfuse", server-driven. */
  kind: string;
  available: boolean;
  required: boolean;
  degraded: boolean;
}

export interface PromptQualitySources {
  degraded: boolean;
  items: PromptQualitySource[];
}

export interface PromptQualityDashboard {
  scope: string;
  scope_id: string;
  /** First calendar day the window covers, YYYY-MM-DD. */
  since: string;
  /** Every version in the range folded together — the headline row. */
  window: PromptQualityVersionMeasures;
  /** The same data split per version, newest first. */
  versions: PromptQualityVersionMeasures[];
  /** Empty when D3 was never scored for this scope. */
  perplexity: PromptQualityPerplexity[];
  data_sources: PromptQualitySources;
}

File: server/internal/handler/prompt_quality.go
package handler

import (
	"encoding/json"
	"net/http"
	"sort"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5/pgtype"

	db "github.com/multica-ai/multica/server/pkg/db/generated"
	"github.com/multica-ai/multica/server/pkg/promptquality"
)

// Prompt quality dashboard read path (RUYI-184, self-evolution phase 2).
//
// It hangs off the same /api/prompt-governance/{scope}/{scopeId} tree as the
// version history, because it answers a question about those versions and
// because the workspace-scope check there is the one this needs: a member of
// the workspace may read, and a scope id belonging to another workspace is a
// 404 before any row is touched (locklessScopeCheck).
//
// Nothing is computed here. The rollup (internal/promptqualityrollup) has
// already folded each day into prompt_quality_daily storing counts rather than
// rates, and pkg/promptquality decides what a count may be rendered as. This
// handler reads rows, groups them by version, and hands each group to
// Combine + Present. The "no data" and "insufficient sample" states arrive
// from that package and are passed through untouched — there is no branch here
// that could substitute a zero.

// promptQualityMaxDays bounds the window. Half a year of daily rows for one
// scope is a few hundred rows; beyond that the timeline stops being readable
// before the query stops being cheap.
const promptQualityMaxDays = 180

const promptQualityDefaultDays = 30

// PromptQualityMeasuresResponse is one version's seven cards plus the span the
// numbers cover. Version 0 is used for the window aggregate across versions.
type PromptQualityMeasuresResponse struct {
	Version  int32                      `json:"version"`
	Days     int                        `json:"days"`
	FirstDay string                     `json:"first_day,omitempty"`
	LastDay  string                     `json:"last_day,omitempty"`
	Runs     int                        `json:"runs"`
	Measures promptquality.Presentation `json:"measures"`
	Reasons  map[string]int             `json:"failure_reasons"`
	Excluded int                        `json:"excluded_failed_runs"`
}

// PromptQualityPerplexityResponse is one D3 score. There is one per runtime
// profile per version and they are never merged: the profiles are scored
// against different assembled documents, so an average would describe a prompt
// no run ever receives (Owner Q17).
type PromptQualityPerplexityResponse struct {
	Version        int32           `json:"version"`
	RuntimeProfile string          `json:"runtime_profile"`
	Band           string          `json:"band"`
	PercentLow     *float64        `json:"percent_low"`
	PercentHigh    *float64        `json:"percent_high"`
	Evidence       json.RawMessage `json:"evidence"`
	Model          string          `json:"model"`
	ScoredAt       string          `json:"scored_at"`
}

// PromptQualityResponse is the dashboard payload.
type PromptQualityResponse struct {
	Scope   string `json:"scope"`
	ScopeID string `json:"scope_id"`
	Since   string `json:"since"`

	// Window is every version in the range folded together, for the headline
	// row. Versions is the same data split per version, which is what the
	// timeline and the cross-version comparison read.
	Window   PromptQualityMeasuresResponse   `json:"window"`
	Versions []PromptQualityMeasuresResponse `json:"versions"`

	// Perplexity is empty when D3 was never scored for this scope. Empty is
	// the honest answer: the scorer writes no row when it refuses, so there is
	// nothing here to distinguish from a neutral band.
	Perplexity []PromptQualityPerplexityResponse `json:"perplexity"`

	// DataSources reports which inputs are switched on. Its only effect on the
	// UI is a footer label — see promptquality.DescribeSources for why no card
	// may depend on it (T3).
	DataSources promptQualitySourcesResponse `json:"data_sources"`
}

type promptQualitySourcesResponse struct {
	Degraded bool                        `json:"degraded"`
	Items    []promptquality.SourceState `json:"items"`
}

// GetPromptQualityDashboard — GET /api/prompt-governance/{scope}/{scopeId}/quality?days=N
func (h *Handler) GetPromptQualityDashboard(w http.ResponseWriter, r *http.Request) {
	scope, ok := parsePromptVersionScope(chi.URLParam(r, "scope"))
	if !ok {
		writeError(w, http.StatusBadRequest, "unknown scope")
		return
	}
	scopeID, ok := parseUUIDOrBadRequest(w, chi.URLParam(r, "scopeId"), "scope_id")
	if !ok {
		return
	}
	workspaceID := parseUUID(h.resolveWorkspaceID(r))

	if !h.locklessScopeCheck(w, r, scope, workspaceID, scopeID) {
		return
	}

	days := promptQualityDefaultDays
	if v := r.URL.Query().Get("days"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 && n <= promptQualityMaxDays {
			days = n
		}
	}
	since := time.Now().UTC().AddDate(0, 0, -(days - 1)).Truncate(24 * time.Hour)

	rows, err := h.Queries.ListPromptQualityDaily(r.Context(), db.ListPromptQualityDailyParams{
		Scope:   string(scope),
		ScopeID: scopeID,
		Since:   pgtype.Date{Time: since, Valid: true},
	})
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to read prompt quality")
		return
	}

	scores, err := h.Queries.ListPromptPerplexityScores(r.Context(), db.ListPromptPerplexityScoresParams{
		Scope:   string(scope),
		ScopeID: scopeID,
	})
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to read prompt quality")
		return
	}

	resp := PromptQualityResponse{
		Scope:       string(scope),
		ScopeID:     uuidToString(scopeID),
		Since:       since.Format("2006-01-02"),
		Versions:    promptQualityByVersion(rows),
		Perplexity:  promptQualityScoresToResponse(scores),
		DataSources: h.promptQualitySources(),
	}

	all := make([]promptquality.Result, 0, len(rows))
	for _, row := range rows {
		all = append(all, promptQualityRowToResult(row))
	}
	resp.Window = promptQualityMeasures(0, rows, all)

	writeJSON(w, http.StatusOK, resp)
}

// promptQualitySources reports the optional inputs' state. The scoring model
// is read from the handler's LLM client rather than the environment so the
// answer matches what the scorer would actually do if asked right now.
func (h *Handler) promptQualitySources() promptQualitySourcesResponse {
	scoring := h.LLM != nil && h.LLM.Enabled()
	sources := promptquality.DescribeSources(scoring)
	return promptQualitySourcesResponse{Degraded: sources.Degraded(), Items: sources.Items}
}

// promptQualityByVersion splits the window per version, newest first. The
// dashboard's timeline is a version axis, not a date axis: a prompt edit is the
// event that can move these numbers, so a version's days belong together even
// when they are not contiguous.
func promptQualityByVersion(rows []db.PromptQualityDaily) []PromptQualityMeasuresResponse {
	grouped := map[int32][]db.PromptQualityDaily{}
	order := []int32{}
	for _, row := range rows {
		if _, seen := grouped[row.Version]; !seen {
			order = append(order, row.Version)
		}
		grouped[row.Version] = append(grouped[row.Version], row)
	}
	sort.Slice(order, func(i, j int) bool { return order[i] > order[j] })

	out := make([]PromptQualityMeasuresResponse, 0, len(order))
	for _, version := range order {
		group := grouped[version]
		results := make([]promptquality.Result, 0, len(group))
		for _, row := range group {
			results = append(results, promptQualityRowToResult(row))
		}
		out = append(out, promptQualityMeasures(version, group, results))
	}
	return out
}

func promptQualityMeasures(version int32, rows []db.PromptQualityDaily, results []promptquality.Result) PromptQualityMeasuresResponse {
	combined := promptquality.Combine(results)
	out := PromptQualityMeasuresResponse{
		Version:  version,
		Days:     len(rows),
		Runs:     combined.FinishedRuns,
		Measures: promptquality.Present(combined),
		Reasons:  combined.FailureReasonCounts,
		Excluded: combined.ExcludedFailedRuns,
	}
	for _, row := range rows {
		if !row.Day.Valid {
			continue
		}
		day := row.Day.Time.UTC().Format("2006-01-02")
		if out.FirstDay == "" || day < out.FirstDay {
			out.FirstDay = day
		}
		if day > out.LastDay {
			out.LastDay = day
		}
	}
	return out
}

// promptQualityRowToResult reverses the rollup's write. Every nullable column
// becomes a nil pointer rather than a zero, which is the whole reason those
// columns are nullable (T5).
func promptQualityRowToResult(row db.PromptQualityDaily) promptquality.Result {
	out := promptquality.Result{
		FinishedRuns:           int(row.FinishedRuns),
		DisciplineCoveredRuns:  int(row.DisciplineCoveredRuns),
		ToolResultsMeasured:    row.ToolResultsMeasured,
		ToolResultsError:       row.ToolResultsError,
		AttemptTotal:           int(row.AttemptTotal),
		RetriedRuns:            int(row.RetriedRuns),
		AttributableFailedRuns: int(row.AttributableFailedRuns),
		ExcludedFailedRuns:     int(row.ExcludedFailedRuns),
		FailureReasonCounts:    map[string]int{},
		FirstPassIssues:        int(row.FirstPassIssues),
		ReviewedIssues:         int(row.ReviewedIssues),
	}
	if row.InjectedTokens.Valid {
		v := row.InjectedTokens.Int64
		out.InjectedTokens = &v
	}
	if row.RunTokensMedian.Valid {
		v := row.RunTokensMedian.Int64
		out.RunTokensMedian = &v
	}
	if v, ok := numericToFloat(row.DisciplineScoreMedian); ok {
		n := int(*v)
		out.DisciplineScoreMedian = &n
	}
	if len(row.FailureReasonCounts) > 0 {
		// A malformed counts blob leaves the map empty rather than failing the
		// request: D6's headline number lives in its own column, and losing the
		// breakdown is not a reason to blank the other six cards.
		_ = json.Unmarshal(row.FailureReasonCounts, &out.FailureReasonCounts)
	}
	return out
}

func promptQualityScoresToResponse(scores []db.PromptPerplexityScore) []PromptQualityPerplexityResponse {
	out := make([]PromptQualityPerplexityResponse, 0, len(scores))
	for _, s := range scores {
		item := PromptQualityPerplexityResponse{
			Version:        s.Version,
			RuntimeProfile: s.RuntimeProfile,
			Band:           s.Band,
			PercentLow:     firstOrNil(numericToFloat(s.PercentLow)),
			PercentHigh:    firstOrNil(numericToFloat(s.PercentHigh)),
			Evidence:       json.RawMessage(s.Evidence),
			Model:          s.Model,
		}
		if len(item.Evidence) == 0 {
			item.Evidence = json.RawMessage("[]")
		}
		if s.ScoredAt.Valid {
			item.ScoredAt = s.ScoredAt.Time.UTC().Format(httpTimeFormat)
		}
		out = append(out, item)
	}
	return out
}

func numericToFloat(n pgtype.Numeric) (*float64, bool) {
	if !n.Valid {
		return nil, false
	}
	f, err := n.Float64Value()
	if err != nil || !f.Valid {
		return nil, false
	}
	v := f.Float64
	return &v, true
}

func firstOrNil(v *float64, _ bool) *float64 { return v }
